client/resource_attribute_restriction.go

```diff
@@ -0,0 +1,413 @@
+				MinItems:     1,
+				Elem:         enumChoiceSchema(),
+			},
+		},
+	}
+}
+
+func enumChoiceSchema() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"value": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotWhiteSpace,
+			},
+			"display_emoji": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"display_colour": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceAttributeRestrictionRead(d *schema.ResourceData, m interface{}) error {
+	c := m.(*Client)
+	attributeID := d.Id()
+
+	attribute, err := c.GetAttributeRestriction(attributeID)
+	if err != nil {
+		return err
+	}
+	if attribute == nil {
+		d.SetId("")
+		return nil
+	}
+
+	if err := d.Set("key", attribute.Key); err != nil {
+		return err
+	}
+	if err := d.Set("description", attribute.Description); err != nil {
+		return err
+	}
+
+	if attribute.Type == "enumattribute" {
+		e, err := parseEnumAttribute(attribute)
+		if err != nil {
+			return err
+		}
+		if err := d.Set("enum", e); err != nil {
+			return err
+		}
+	}
+
+	if attribute.Type == "freetextattribute" {
+		ft, err := parseNonEnumAttribute(attribute)
+		if err != nil {
+			return err
+		}
+		if err := d.Set("freetext", ft); err != nil {
+			return err
+		}
+	}
+
+	if attribute.Type == "booleanattribute" {
+		b, err := parseNonEnumAttribute(attribute)
+		if err != nil {
+			return err
+		}
+		if err := d.Set("boolean", b); err != nil {
+			return err
+		}
+	}
+
+	if attribute.Type == "integerattribute" {
+		i, err := parseNonEnumAttribute(attribute)
+		if err != nil {
+			return err
+		}
+		if err := d.Set("integer", i); err != nil {
+			return err
+		}
+	}
+
+	if err := d.Set("applies_to", mapTargetsToFrontend(attribute.AppliesTo)); err != nil {
+		return err
+	}
+	return err
+}
+
+func resourceAttributeRestrictionCreate(d *schema.ResourceData, m interface{}) error {
+	c := m.(*Client)
+	attribute, err := composeAttribute(d)
+	if attribute == nil || err != nil {
+		return err
+	}
+
+	a, err := c.CreateAttributeRestriction(*attribute)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(strconv.Itoa(a.ID))
+	return err
+}
+
+func resourceAttributeRestrictionUpdate(d *schema.ResourceData, m interface{}) error {
+	c := m.(*Client)
+	attributeID := d.Id()
+	attribute, err := composeAttribute(d)
+	if attribute == nil || err != nil {
+		return err
+	}
+
+	err = c.UpdateAttributeRestriction(attributeID, *attribute)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceAttributeRestrictionDelete(d *schema.ResourceData, m interface{}) error {
+	c := m.(*Client)
+	attributeID := d.Id()
+
+	err := c.DeleteAttributeRestriction(attributeID)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func composeAttribute(d *schema.ResourceData) (*AttributeRestriction, error) {
+	appliesTo := mapTargetsToBackend(expandStringList(d.Get("applies_to").([]interface{})))
```

Comment:
This type assertion won't work. The value of a TypeSet is `*schema.Set`.

Oddly, setting a TypeSet with a slice does work, because they use a whole lot of reflection.

Reply:
<img width="1087" alt="image" src="https://user-images.githubusercontent.com/5205457/170975857-5e8f72de-f734-4cf4-84a4-8332eacca292.png">